sdk/keyvault/keyvault-secrets/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import { TokenCredential, AccessToken } from "@azure/core-http";
+import { BaseRequestPolicy, RequestPolicy, RequestPolicyOptions, RequestPolicyFactory } from "@azure/core-http";
+import { Constants } from "@azure/core-http";
+import { HttpOperationResponse } from "@azure/core-http";
+import { HttpHeaders } from "@azure/core-http";
+import { WebResource } from "@azure/core-http";
+
+export const TokenRefreshBufferMs = 2 * 60 * 1000; // 2 Minutes
+
+/**
+ * Creates a new ChallengeBasedAuthenticationPolicy factory.
+ *
+ * @param credential The TokenCredential implementation that can supply the challenge token.
+ */
+export function challengeBasedAuthenticationPolicy(credential: TokenCredential): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+      return new ChallengeBasedAuthenticationPolicy(nextPolicy, options, credential);
+    }
+  };
+}
+
+export class AuthenticationChallenge {
+  constructor(public scopes: string[] | string) { }
+}
+
+/**
+ *
+ * Provides a RequestPolicy that can request a token from a TokenCredential
+ * implementation and then apply it to the Authorization header of a request
+ * as a Bearer token.
+ *
+ */
+export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
+  private cachedToken: AccessToken | undefined = undefined;
+  private challenge: AuthenticationChallenge | undefined = undefined;
+
+  /**
+   * Creates a new ChallengeBasedAuthenticationPolicy object.
+   *
+   * @param nextPolicy The next RequestPolicy in the request pipeline.
+   * @param options Options for this RequestPolicy.
+   * @param credential The TokenCredential implementation that can supply the bearer token.
+   * @param scopes The scopes for which the bearer token applies.
+   */
+  constructor(
+    nextPolicy: RequestPolicy,
+    options: RequestPolicyOptions,
+    private credential: TokenCredential,
+    private refreshOn: number = Date.now(),
+  ) {
+    super(nextPolicy, options);
+  }
+
+  private parseWWWAuthenticate(www_authenticate: string): string {
```

Comment:
This code seems arbitrary. Could we provide examples of the input that is expected in the form of comments? Or a link to an external reference, if available.

Reply:
I could comment this, good call. Right now it assumes you know what the WWW-Authenticate string looks like